adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -924,55 +1134,51 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] extends Logging {
    *
    * @see rightOuterBroadcastRegionJoin
    */
-  def rightOuterBroadcastRegionJoinAgainst[X](
-    broadcastTree: Broadcast[IntervalArray[ReferenceRegion, X]])(
-      implicit tTag: ClassTag[T], xTag: ClassTag[X]): GenericGenomicRDD[(Option[X], T)] = RightOuterBroadcastJoin.time {
+  def rightOuterBroadcastRegionJoinAgainst[X, Y <: Product, Z <: GenomicDataset[X, Y, Z]](
+    broadcast: GenomicBroadcast[X, Y, Z])(
+      implicit tTag: ClassTag[T], xTag: ClassTag[X],
+      oyuTag: TypeTag[(Option[Y], U)]): GenericGenomicDataset[(Option[X], T), (Option[Y], U)] = RightOuterBroadcastJoin.time {
 
     // key the RDDs and join
-    GenericGenomicRDD[(Option[X], T)](RightOuterTreeRegionJoin[X, T]().join(
-      broadcastTree,
+    RDDBoundGenericGenomicDataset[(Option[X], T), (Option[Y], U)](RightOuterTreeRegionJoin[X, T]().join(
+      broadcast.broadcastTree,
       flattenRddByRegions()),
-      sequences,
+      sequences ++ broadcast.backingDataset.sequences,
       kv => {
-        getReferenceRegions(kv._2) // FIXME
-      })
+        Seq(kv._1.map(v => broadcast.backingDataset.getReferenceRegions(v))).flatten.flatten ++
+          getReferenceRegions(kv._2)
+      },
+      kv => (kv._1.map(broadcast.backingDataset.productFn), productFn(kv._2)),
+      kv => (kv._1.map(broadcast.backingDataset.unproductFn), unproductFn(kv._2)))
+
   }
 
   /**
-   * Performs a broadcast inner join between this RDD and another RDD.
+   * Performs a broadcast right outer join between this RDD and another RDD.
    *
    * In a broadcast join, the left RDD (this RDD) is collected to the driver,
    * and broadcast to all the nodes in the cluster. The key equality function
    * used for this join is the reference region overlap function. Since this
-   * is an inner join, all values who do not overlap a value from the other
-   * RDD are dropped.
+   * is a right outer join, all values in the left RDD that do not overlap a
+   * value from the right RDD are dropped. If a value from the right RDD does
+   * not overlap any values in the left RDD, it will be paired with a `None`
+   * in the product of the join.
    *
    * @param genomicRdd The right RDD in the join.
-   * @param flankSize Sets a flankSize for the distance between elements to be
-   *   joined. If set to 0, an overlap is required to join two elements.
    * @return Returns a new genomic RDD containing all pairs of keys that
-   *   overlapped in the genomic coordinate space.
+   *   overlapped in the genomic coordinate space, and all keys from the
+   *   right RDD that did not overlap a key in the left RDD.
    *
-   * @see broadcastRegionJoinAgainstAndGroupByRight
+   * @see rightOuterBroadcastRegionJoin
    */
-  def broadcastRegionJoinAndGroupByRight[X, Y <: GenomicRDD[X, Y]](
-    genomicRdd: GenomicRDD[X, Y],
-    flankSize: Long)(
+  def rightOuterBroadcastRegionJoin[X, Y <: Product, Z <: GenomicDataset[X, Y, Z]](
+    genomicRdd: GenomicDataset[X, Y, Z])(
```

Comment:
genomicRdd &rarr; genomicDataset

Reply:
How much do you want me to go through and patch these? For now, I've left them as is (genomicRdd). I'd rather make a cleanup pass through later.